src/io/udp_frame.rs

```diff
@@ -0,0 +1,251 @@
+    /// `Ok(None)` is returned. This indicates to the `Framed` instance that
+    /// it needs to read some more bytes before calling this method again.
+    /// In such a case, it is decode's responsibility to copy the data
+    /// into their own internal buffer for future use.
+    ///
+    /// Finally, if the bytes in the buffer are malformed then an error is
+    /// returned indicating why. This informs `Framed` that the stream is now
+    /// corrupt and should be terminated.
+    ///
+    fn decode(&mut self, src: &SocketAddr, buf: &[u8]) -> Result<Option<Self::In>, io::Error>;
+}
+
+/// A unified `Stream` and `Sink` interface to an underlying `Io` object, using
+/// the `CodecUdp` trait to encode and decode frames.
+///
+/// You can acquire a `Framed` instance by using the `Io::framed` adapter.
+pub struct FramedUdp<C> {
+    socket: UdpSocket,
+    codec: C,
+    out_addr : Option<SocketAddr>,
+    rd: Vec<u8>,
+    wr: Vec<u8>,
+}
+
+impl<C : CodecUdp> Stream for FramedUdp<C> {
+    type Item = C::In;
+    type Error = io::Error;
+
+    fn poll(&mut self) -> Poll<Option<C::In>, io::Error> {
+        loop {
+
+            let ret = self.socket.recv_from(self.rd.as_mut_slice());
+            match ret {
+                Ok((n, addr)) => { 
+                    trace!("read {} bytes", n);
+                    trace!("attempting to decode a frame");
+                    if let Some(frame) = try!(self.codec.decode(&addr, & self.rd[.. n])) {
+                        trace!("frame decoded from buffer");
+                        return Ok(Async::Ready(Some(frame)));
+                    }
+                }
+                Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
+                    return Ok(Async::NotReady)
+                }
+                Err(e) => return Err(e),
+            }
+        }
+    }
+}
+
+impl<C : CodecUdp> Sink for FramedUdp<C> {
+    type SinkItem = C::Out;
+    type SinkError = io::Error;
+
+    fn start_send(&mut self, item: C::Out) -> StartSend<C::Out, io::Error> {
+        if self.wr.len() > 0 {
+            try!(self.poll_complete());
+            if self.wr.len() > 0 {
+                return Ok(AsyncSink::NotReady(item));
+            }
+        }
+
+        self.out_addr = Some(self.codec.encode(&item, &mut self.wr));
+        Ok(AsyncSink::Ready)
+    }
+
+    fn poll_complete(&mut self) -> Poll<(), io::Error> {
+        trace!("flushing framed transport");
+
+        while !self.wr.is_empty() {
+            if let Some(outaddr) = self.out_addr {
+                let remaining = self.wr.len();
+                trace!("writing; remaining={}", remaining);
+                let n = try_nb!(self.socket.send_to(self.wr.as_slice(), &outaddr));
+                trace!("written {}", n);
+                self.wr.clear();
+                self.out_addr = None;
+                if n != remaining {
+                    return Err(io::Error::new(io::ErrorKind::WriteZero,
```

Comment:
I'd be fine with just an `Other` error kind here for now as `WriteZero` isn't *quite* what happened.